client/fingerprint/network_unix.go

```diff
@@ -152,99 +144,82 @@ func (f *NetworkFingerprint) linkSpeedEthtool(path, device string) int {
+		return "", errors.New(fmt.Sprintf("Interface %s has no IP address", intf.Name))
+	}
+	var ipV4 net.IP
+	for _, addr := range addrs {
+		var ip net.IP
+		switch v := (addr).(type) {
+		case *net.IPNet:
+			ip = v.IP
+		case *net.IPAddr:
+			ip = v.IP
 		}
-
-		addrs, err := i.Addrs()
-		if err != nil {
-			return "", errors.New("could not retrieve interface IP addresses")
-		}
-
-		for _, a := range addrs {
-			switch v := a.(type) {
-			case *net.IPNet:
-				if v.IP.To4() != nil {
-					ip = v.IP.String()
-				}
-			case *net.IPAddr:
-				if v.IP.To4() != nil {
-					ip = v.IP.String()
-				}
-			}
+		if ip.To4() != nil {
+			ipV4 = ip
+			break
 		}
 	}
 
-	if net.ParseIP(ip) == nil {
-		return "", errors.New(fmt.Sprintf("could not parse IP address `%s`", ip))
+	if ipV4 == nil {
+		return "", errors.New(fmt.Sprintf("Couldn't parse IP address for interface %s with addr %s", intf.Name))
 	}
+	return ipV4.String(), nil
 
-	return ip, nil
 }
 
-// ifConfig returns the IP Address for this node according to ifConfig, for the
-// specified device.
-func (f *NetworkFingerprint) ifConfig(device string) string {
-	ifConfigPath, _ := exec.LookPath("ifconfig")
-	if ifConfigPath == "" {
-		f.logger.Println("[WARN] fingerprint.network: ifconfig not found")
-		return ""
-	}
+// Checks if the device is marked UP by the operator
+func (f *NetworkFingerprint) isDeviceEnabled(intf *net.Interface) bool {
+	return intf.Flags&net.FlagUp != 0
+}
 
-	outBytes, err := exec.Command(ifConfigPath, device).Output()
-	if err != nil {
-		f.logger.Printf("[WARN] fingerprint.network: Error calling ifconfig (%s %s): %v", ifConfigPath, device, err)
-		return ""
+// Checks if the device has any IP address configured
+func (f *NetworkFingerprint) deviceHasIpAddress(intf *net.Interface) bool {
+	if _, err := f.ipAddress(intf); err == nil {
+		return true
 	}
+	return false
+}
 
-	// Parse out the IP address returned from ifconfig for this device
-	// Tested on Ubuntu, the matching part of ifconfig output for eth0 is like
-	// so:
-	//   inet addr:10.0.2.15  Bcast:10.0.2.255  Mask:255.255.255.0
-	// For OS X and en0, we have:
-	//  inet 192.168.0.7 netmask 0xffffff00 broadcast 192.168.0.255
-	output := strings.TrimSpace(string(outBytes))
-
-	// re is a regular expression, which can vary based on the OS
-	var re *regexp.Regexp
+func (n *NetworkFingerprint) isDeviceLoopBackOrPointToPoint(intf *net.Interface) bool {
+	return intf.Flags&(net.FlagLoopback|net.FlagPointToPoint) == 0
+}
 
-	if "darwin" == runtime.GOOS {
-		re = regexp.MustCompile("inet [0-9].+")
-	} else {
-		re = regexp.MustCompile("inet addr:[0-9].+")
+// Returns interfaces which are routable and marked as UP
+// Tries to get the specific interface if the user has specified name
+func (f *NetworkFingerprint) findInterfaces(deviceName string) ([]*net.Interface, error) {
+	var (
+		interfaces []*net.Interface
+		err        error
+	)
+	if deviceName != "" {
+		if intf, err := net.InterfaceByName(deviceName); err == nil {
```

Comment:
This logic seems like it should be in a separate method. Shouldn't we also check if the passed device's interface is Up and has an IP like the checks below